In [1]:
import csv

In [2]:
central_stmt = "{} mean is {} and median is {}."
deviation_stmt = "{} variance is {} and standard deviation is {}."
co_measures_stmt = "{} and {} covariance is {} and correlation is {}."

In [3]:
with open('zadanie_korelacja_source.csv') as source_file:
    source_data = csv.reader(source_file, delimiter=',')
    
    headers = next(source_data)
    
    data_collection = [(float(row[1]), float(row[2])) for row in source_data]
    
    x_label, y_label = headers[1], headers[2]
    
data_collection

[(1.3, 76.84878),
 (1.95, 81.40732),
 (2.466, 77.57895),
 (1.6, 74.16341),
 (2.952, 68.84907),
 (1.801, 73.88595),
 (1.34, 75.26829),
 (1.34, 76.26098),
 (1.63, 80.57244),
 (2.326, 65.46259),
 (2.436, 67.5482),
 (6.563, 55.02451),
 (1.283, 76.2799),
 (1.43, 82.29024),
 (1.988, 69.80695),
 (1.61, 81.40112),
 (1.921, 82.19756),
 (2.4, 74.22683),
 (1.495, 80.12888),
 (4.705, 45.55095)]

In [4]:
def mean(input_list, column):
    collection = [row[column] for row in input_list]
    
    result = sum(collection) / len(collection)
    
    return result

mean(data_collection, 1)

73.237646

In [5]:
def median(input_list, column):
    collection = [row[column] for row in input_list]
    
    collection.sort()    
    input_parity = len(collection) % 2
    sample_size = len(collection)
    
    if input_parity == 0:
        a = int(sample_size / 2 - 1)
        b = int(sample_size / 2)
        result = (collection[a] + collection[b]) / 2
    else:
        c = round(sample_size / 2)
        result = collection[c]
        
    return result

median(data_collection, 1)

75.764635

In [27]:
def mode(input_list, column):
    collection = [row[column] for row in input_list]
    frequencies = [(collection.count(x), x) for x in collection]
    frequencies = list(dict.fromkeys(frequencies))
    modes = [i for i in frequencies if i[0] > 1]
    
    return modes

mode(data_collection, 1)

[]

In [7]:
def variance(input_list, column):
    collection = [row[column] for row in input_list]
    
    sample_mean = sum(collection) / len(collection)
    deviances = [(i - sample_mean) ** 2 for i in collection]
    
    result = sum(deviances) / (len(collection) - 1)
    
    return result

variance(data_collection, 1)

88.19411239750946

In [8]:
def st_dev(input_list, column):
        
    result = variance(input_list, column) ** (0.5)
    
    return result

st_dev(data_collection, 1)

9.39117204599668

In [9]:
def covariance(input_list):
    x_mean = mean(input_list, 0)
    y_mean = mean(input_list, 1)
    deviances = [(i[0] - x_mean) * (i[1] - y_mean) for i in input_list]
    
    result = sum(deviances) / (len(deviances) - 1)
    
    return result

covariance(data_collection)

-9.943704446631578

In [10]:
def correlation(input_list):
    
    result = covariance(input_list) / (st_dev(input_list, 0) * st_dev(input_list, 1))
    
    return result

correlation(data_collection)

-0.8204930122649574

In [11]:
def conf_interval(input_list, column, pop_st_dev, level):
    zets = {99: 2.576, 98: 2.326, 95: 1.96, 90: 1.645}
    z = zets[level]
    
    collection = [row[column] for row in input_list]
    
    sample_mean = sum(collection) / len(collection)    
    factor = z * (pop_st_dev / (len(data_collection) ** (1/2)))
    
    result = ((sample_mean - factor), (sample_mean + factor))
    
    return result

conf_interval(data_collection, 1, 9.09, 95)

(69.25377784856727, 77.22151415143273)

In [12]:
y_mean = mean(data_collection, 1)
y_median = median(data_collection, 1)

y_variance = variance(data_collection, 1)
y_st_dev = st_dev(data_collection, 1)

xy_covariance = covariance(data_collection)
xy_correlation = correlation(data_collection)

given_conf_int = conf_interval(data_collection, 1, 9.09, 95)

In [13]:
print(central_stmt.format(y_label, "{:.2f}".format(y_mean),
                          "{:.2f}".format(y_median)))                      
print(deviation_stmt.format(y_label, "{:.2f}".format(y_variance),
                            "{:.2f}".format(y_st_dev)))
print(co_measures_stmt.format(x_label, y_label, "{:.2f}".format(xy_covariance),
                              "{:.2f}".format(xy_correlation)))
print("{} 95% confidence interval for its mean is from {} to {}.".format(y_label,
                                                                     "{:.2f}".format(given_conf_int[0]),
                                                                     "{:.2f}".format(given_conf_int[1])))

Life expectancy at birth mean is 73.24 and median is 75.76.
Life expectancy at birth variance is 88.19 and standard deviation is 9.39.
Fertility rate and Life expectancy at birth covariance is -9.94 and correlation is -0.82.
Life expectancy at birth 95% confidence interval for its mean is from 69.25 to 77.22.


# Statistics for data provided

Life expectancy at birth mean is 73.24 and median is 75.76.
Life expectancy at birth collection has no modes.
Life expectancy at birth 95% confidence interval for its mean is from 69.25 to 77.22.
Life expectancy at birth variance is 88.19 and standard deviation is 9.39.
Fertility rate and Life expectancy at birth covariance is -9.94 and correlation is -0.82.

Considering above figures, very strong negative correlation between Fertility rate and Life expectancy at birth has been observed. Such a correlation could be explained by a level of well-being in the countries observed, but further investigation to confirm this hypothesis is needed.
